src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/OpenSslX509ChainProcessor.cs

```diff
@@ -921,7 +921,20 @@ private static ArraySegment<char> Base64UrlEncode(ReadOnlySpan<byte> input)
                         statusBuilder ??= new List<X509ChainStatus>();
                         overallStatus ??= new List<X509ChainStatus>();
 
-                        AddElementStatus(elementErrors.Value, statusBuilder, overallStatus);
+                        bool hadSignatureNotValid = overallHasNotSignatureValid;
+                        AddElementStatus(elementErrors.Value, statusBuilder, overallStatus, ref overallHasNotSignatureValid);
+
+                        // Clear NotSignatureValid for the last element when overall chain is not PartialChain or UntrustedRoot.
+                        bool isLastElement = i == chainSize - 1;
+                        if (isLastElement && !hadSignatureNotValid && overallHasNotSignatureValid)
+                        {
+                            if (!ContainsStatus(overallStatus, X509ChainStatusFlags.PartialChain) &&
+                                !ContainsStatus(overallStatus, X509ChainStatusFlags.UntrustedRoot))
```

Comment:
This is O(overallStatus.Count), and it's contained inside a loop from 0 to chainSize.  Are the lengths involved here always small enough that the O(overallStatus.Count * chainSize) isn't an issue?

Reply:
We're only performing the innerloop for the last element of the outerloop.

Reply:
There are only 26 status codes, many of which are mutually exclusive.

The average chain element has no problems, so no codes.  After that it'll have an average of 2-ish.

If you wanted to craft a really degenerate case you might be able to get it up to 17.

1. NotTimeValid
2. NotTimeNested
3. NotSignatureValid
4. NotValidForUsage
5. UntrustedRoot
6. RevocationStatusUnknown
7. InvalidExtension
8. InvalidPolicyConstraints
9. InvalidBasicConstraints
10. InvalidNameConstraints
11. HasNotSupportedNameConstraint
12. HasNotDefinedNameConstraint
13. HasNotPermittedNameConstraint
14. HasExcludedNameConstraint
15. OfflineRevocation
16. NoIssuanceChainPolicy
17. HasNotSupportedCriticalExtension